oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java

```diff
@@ -582,6 +576,53 @@ void executeIfNew(Executor executor) {
     }
   }
 
+  @VisibleForTesting
+  class RefreshTaskListener implements Runnable {
+    private ListenableFutureTask<OAuthValue> task;
+
+    RefreshTaskListener(ListenableFutureTask<OAuthValue> task) {
+      this.task = task;
+    }
+
+    @Override
+    public void run() {
+      finishRefreshAsync(task);
+    }
+  }
+
+  class RefreshTask extends AbstractFuture<OAuthValue> implements Runnable {
+    private final ListenableFutureTask<OAuthValue> task;
+    private final RefreshTaskListener listener;
+
+    RefreshTask(ListenableFutureTask<OAuthValue> task, RefreshTaskListener listener) {
+      this.task = task;
+      this.listener = listener;
+
+      // Update Credential state first
+      task.addListener(listener, MoreExecutors.directExecutor());
```

Comment:
I think you should either in-line both callbacks or move all the callback logic into the listener class.

Reply:
In order to inject a sleep to exercise the race condition I wrote it like this. Agreed the logic could be better but I am not very familiar with Java

Reply:
That said I would happily accept a cleaner approach!